### ISSUE TITLE
Define Behavior of `ThreadRunModel#advanceFrom()` In Unhappy Path

### ISSUE THREAD

**Issue by coltmcnealy-lh** [littlehorse-enterprises/littlehorse#656]
# Background

Currently, there are two versions of `ThreadRunModel#advanceFrom()`:

1. `public void advanceFrom(NodeModel curNode, VariableValueModel output)`
2. `public void advanceFrom(NodeModel curNode)`

The first one takes in the output of the `NodeRun` and the second one does not. Method 2) simply calls method 1) with `null` as the output. For context, the output of a `NodeRun` might be:
* The `content` of an `ExternalEvent` for an `ExternalEventNodeRun`
* The `result` of a `TaskRun` for a `TaskNodeRun`
* The `results` of a `UserTaskRun` for a `UserTaskNodeRun`
* And so on.

# Current Behavior Scenarios

There are three cases in which some flavor of `advanceFrom()` is called:
1. Happy Path: `NodeRun` completes while the `ThreadRun` is `RUNNING`
2. The `NodeRun` completes while the `ThreadRun` is `HALTING`; and _then_ the `ThreadRun` moves to `RUNNING`.
3. The `NodeRun` fails, and then the `FailureHandler` completes successfully.

## Happy Path
The Happy Path is quite simple: the `ThreadRun` is `RUNNING` and the `NodeRun` completes without incident. In this case, what happens is:
- Each `Edge` in the `Node.outgoing_edges` is evaluated in order.
- The first matching `Edge` is chosen.
- The `Edge` performs its `VariableMutation`s
- The `ThreadRun` advances to the `Node` specified by the `Edge`'s `sink_node_name`.

## Completed while `HALTING`
Let's imagine the following scenario:
1. `ThreadRun` arrives at an `ExternalEventNodeRun`.
2. `ThreadRun` moves to `HALTED` state.
3. The `ExternalEvent` arrives and is matched with the `ExternalEventNodeRun`.
4. The `ThreadRun` is moved to `RUNNING` again.

We have *highly incorrect behavior* here. What happens is that after step 4), when the `ThreadRun` is moved to `RUNNING`, we call `advanceFrom(Date time)` and _ignore_ the ouput of the `ExternalEvent`. This means that any `VariableMutation`s that rely on the `output` (in this case, the `ExternalEvent`) will cause a `NullPointerException` in the `mutateVariables()` method.

### What Should Happen

What _should_ happen here is clear. We should look up the output of the successfully completed `NodeRun` and use that output when calling the `advanceFrom()` method. In order to do this, we should probably introduce a new method on the `SubNodeRunModel`

```
public VariableValueModel getOutput();
```
Each implementation of `SubNodeRunModel` will implement that method. For example, the `ExternalEventNodeRunModel` will just look up the `ExternalEvent` and return the `.getContent()`.

## Advancing after Failure Handler

Finally, let's consider what happens in the scenario where there is a Failure Handler.
1. A `ThreadRun` arrives at a `TaskNode`, and the `TaskRun` fails.
2. There is a `FailureHandlerDef` prescribed for the `Failure` that was thrown.
3. The `FailureHandler` `ThreadRun` completes successfully.

After step 3), we _also_ call `ThreadRunModel#advance(Date time)` without passing in the `VariableValueModel output`. This leads to similar undefined behavior in the `ThreadRunModel#mutateVariables()` method.

### What Should Happen

In order to understand what _should_ happen in this scenario, we should first look at the semantics of the output of a `NodeRun`. Consider the following `WfSpec` code in the Java SDK:

```
WfRunVariable salesforceRecordId = wf.addVariable("salesforce-record-id", STR);
NodeOutput taskOutput = wf.execute("create-salesforce-record");

wf.handleAnyFailure(taskOutput, this::someFailureThread);
wf.mutate(salesforceRecordId, VariableMutationType.ASSIGN, taskOutput);
```
In this code, we call a `TaskRun` that returns a `STR` that contains a salesforce record ID. In the outgoing edges from that node, we mutate the `salesforce-record-id` variable and assign it the output of the `TaskRun`.

If the task fails, and the Failure Handler successfully executes, we _still_ don't have a valid `salesforce-record-id`. So we can't execute the `VariableMutation`. And further steps in the `WfRun` will depend on the`salesforce-record-id`.

*I propose that we enable a `ThreadRun` to return an `output`, and use the `output` of the `FailureHandler` as the `output` of the `NodeRun`*.

**Comment by coltmcnealy-lh** [comment]
cc @eduwercamacaro this is a big one...Definitely needed for `1.0.0` but not `0.7.2`.

**Comment by eduwercamacaro** [comment]
I do agree on adding a `public VariableValueModel getOutput();` method for SubNodeRun implementations, but I prefer adding this at NodeRun level. If we add at NodeRun level that will complicate the implementation, so lets add it at SubNodeRun level. In the other hand I propose this method signature instead:

`public Optional<VariableValueModel> getOutput();`

As far I understand, the output is optional ;)

**Comment by eduwercamacaro** [comment]
I don't think we should enable `ThreadRun` to return an output. It seems incorrect. 

I believe that in this example:


```
WfRunVariable salesforceRecordId = wf.addVariable("salesforce-record-id", STR);
NodeOutput taskOutput = wf.execute("create-salesforce-record");

wf.handleAnyFailure(taskOutput, this::someFailureThread);
wf.mutate(salesforceRecordId, VariableMutationType.ASSIGN, taskOutput);
```

After line number 4, salesforceRecordId should be null if `create-salesforce-record` fails.

**Comment by coltmcnealy-lh** [comment]
Agreed.

Colt McNealy

*Founder, LittleHorse.dev*

**Comment by coltmcnealy-lh** [comment]
> I don't think we should enable `ThreadRun` to return an output. It seems incorrect.
> 
> I believe that in this example:
> 
> ```
> WfRunVariable salesforceRecordId = wf.addVariable("salesforce-record-id", STR);
> NodeOutput taskOutput = wf.execute("create-salesforce-record");
> 
> wf.handleAnyFailure(taskOutput, this::someFailureThread);
> wf.mutate(salesforceRecordId, VariableMutationType.ASSIGN, taskOutput);
> ```
> 
> After line number 4, salesforceRecordId should be null if `create-salesforce-record` fails.

We are definitely going to enable `ThreadRun`'s to return an Output (this will be done on the `EXIT` Node) as we need it for some other features on our roadmap.

As for whether we want to make the `FailureHandler` `ThreadRun`'s output be used as the `NodeOutput`, we can have that discussion in a separate ticket. For now we will just add some end-to-end tests to ensure that we get the `VariableValue` treated as `VALUE_NOT_SET` and ensure that the server doesn't have any NPE's...

Anyways, I want to design `ThreadRun` Output's with the other features in mind rather than just this feature. So this ticket will not have `ThreadRun` output in scope.

**Comment by coltmcnealy-lh** [comment]
I have done some more thinking about what should happen with `VariableMutation`s in the case of a `Failure` and `FailureHandler`. I think I am starting to agree with @eduwercamacaro 's previous suggestion:
>After line number 4, salesforceRecordId should be null if create-salesforce-record fails.

This is because, consider the following Java code:
```
String myVar = null;
try {
  myVar = someMethodThatThrowsAnException();
} catch (Exception exn) {
  // *X*
}
System.out.println(myVar);
```
The above code will print out `null`. That is the same behavior as what Edu is suggesting.

In Java, this gives the user the flexibility to do any of the following:

1. Put some logic in the exception handler (where I drew the `*X*`) which sets the value of `myVar`.
2. Throw another exception inside the exception handler if it's not possible to continue with `myVar` being null.
3. Simply allow the exception handler to continue, and let `myVar` remain `null` if that is acceptable.

My proposal is as follows:
_If there is a `Failure` on a `NodeRun`, then we do NOT execute the `VariableMutation`s._

This provides equivalent behavior to the way Java exception handlers work, as the user can choose one of the following options:
1. Put logic in the `FailureHandler` `ThreadSpec` that mutates the parent's variable.
2. Throw another `Failure` inside the `FailureHandler` if you can't continue with `salesforce-id` being `NULL`
3. Allow the `FailureHandler` to just succeed, and `salesforce-id` remains `NULL`

**Comment by eduwercamacaro** [comment]
I agree with your last proposal :+1:

**Comment by coltmcnealy-lh** [comment]
Closed by #665
